base_layer/core/src/transactions/transaction_components/encrypted_openings.rs

```diff
@@ -0,0 +1,289 @@
+
+//! Encrypted openings using the the extended-nonce variant XChaCha20-Poly1305 encryption with secure random nonce.
+
+use std::mem::size_of;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use chacha20poly1305::{
+    aead::{Aead, Error, Payload},
+    KeyInit,
+    Tag,
+    XChaCha20Poly1305,
+    XNonce,
+};
+use digest::{generic_array::GenericArray, FixedOutput};
+use rand::{rngs::OsRng, RngCore};
+use serde::{Deserialize, Serialize};
+use tari_common_types::types::{Commitment, PrivateKey};
+use tari_crypto::{hash::blake2::Blake256, hashing::DomainSeparatedHasher};
+use tari_utilities::{
+    hex::{from_hex, to_hex, Hex, HexError},
+    safe_array::SafeArray,
+    ByteArray,
+    ByteArrayError,
+};
+use thiserror::Error;
+use zeroize::Zeroize;
+
+use super::EncryptedOpeningsKey;
+use crate::transactions::{tari_amount::MicroTari, TransactionSecureNonceKdfDomain};
+
+const VALUE_SIZE: usize = size_of::<u64>(); // 8 bytes
+const KEY_SIZE: usize = size_of::<PrivateKey>(); // 32 bytes
+const SIZE: usize = VALUE_SIZE + KEY_SIZE + size_of::<Tag>() + size_of::<XNonce>(); // 80 bytes
+const BORSH_64: usize = 64;
+const BORSH_X: usize = SIZE - BORSH_64; // 16 bytes
+
+/// Encrypted openings for the extended-nonce variant XChaCha20-Poly1305 encryption
+/// Borsh schema only accept array sizes 0 - 32, 64, 65, 128, 256, 512, 1024 and 2048
+#[derive(
+    Debug, Copy, Clone, Deserialize, Serialize, PartialEq, Eq, Hash, BorshSerialize, BorshDeserialize, Zeroize,
+)]
+pub struct EncryptedOpenings {
+    #[serde(with = "tari_utilities::serde::hex")]
+    data_1: [u8; BORSH_64],
+    #[serde(with = "tari_utilities::serde::hex")]
+    data_2: [u8; BORSH_X],
+}
+
+impl EncryptedOpenings {
+    const TAG: &'static [u8] = b"TARI_AAD_VALUE_AND_MASK_EXTEND_NONCE_VARIANT";
+
+    /// Encrypt the value and mask (with fixed length) using XChaCha20-Poly1305 with a secure random nonce
+    /// Notes: - This implementation does not require or assume any uniqueness for `encryption_key` or `commitment`
+    ///        - With the use of a secure random nonce, there's no added security benefit in using the commitment in the
+    ///          internal key derivation; but it binds the encrypted openings to the commitment
+    pub fn encrypt_openings(
+        encryption_key: &PrivateKey,
+        commitment: &Commitment,
+        value: MicroTari,
+        mask: &PrivateKey,
+    ) -> Result<EncryptedOpenings, EncryptedOpeningsError> {
+        let mut openings = value.as_u64().to_le_bytes().to_vec();
+        openings.append(&mut mask.to_vec());
+        let aead_payload = Payload {
+            msg: openings.as_slice(),
+            aad: Self::TAG,
+        };
+
+        // Produce a secure random nonce
+        let mut nonce = [0u8; size_of::<XNonce>()];
+        OsRng.fill_bytes(&mut nonce);
+        let nonce_ga = XNonce::from_slice(&nonce);
+
+        let aead_key = kdf_aead(encryption_key, commitment);
```

Comment:
Depending on how [this draft RFC](https://github.com/tari-project/rfcs/pull/98) evolves, it may be more efficient to nix this KDF, and offload key generation to the caller. This would be particularly relevant for cases where `encryption_key` is fixed across multiple outputs, and where the addition of an internal KDF would add a lot of overhead.

Reply:
This sounds good, but I think that's better for a next PR

Reply:
Agreed for a nex tPR

Reply:
Noted in a new [issue](https://github.com/tari-project/tari/issues/5393).